### ISSUE TITLE
Using conference-app.zip application with Dialogic JSR 309 Connector and XMS error in a play: cant perform play no MIXER or NC found associated to this IVR container

### ISSUE THREAD

**Issue by RomanDS** [RestComm/RestComm-Core#788]
Issue seems to be related to the fact that conference application issues playAsync before the join between MG and NC is completed. The Join (unlike JoinInitiate) is blocking function so it should hold until it completes. The question seems to be around a play which seems to be executed on a different thread. 
Q: How are the two synchronized to make sure play request does not get called before Join completes?

Bellow is an extract with timestamps of the analysis of the logs which are attached:

[server.zip](https://github.com/RestComm/RestComm-Core/files/89217/server.zip)

09:58:30,276 ERROR [com.vendor.dialogic.javax.media.mscontrol.mediagroup.DlgcPlayer - 117] 
(MSS-Executor-Thread-9) 
DlgcPlayer::play cant perform play no MIXER or NC found associated to this IVR containe

XNC-2b99d4c75f1f4 join with XMG-2b99d49565235

MSS-Executor-Thread-0 (join)
Timeline:

09:58:30,247    line 6147 - XNC-2b99d4c75f1f4 Enter join with XMG-2b99d49565235 - Executor-Thread-0

09:58:30,261      line 6283 - request to playAsync Proxy (XMG-2b99d49565235) MSS-Executor-Thread-9

09:58:30,265      line 6365 -  DlgcProxyHelper joinAsyncWithValidation :326 - Executor-Thread-0

09:58:30,276      line 6547 - Play Starts Error reported - MSS-Executor-Thread-9

09:58:30,323      line 7068 - Join returns complete - MSS-Executor-Thread-0

09:58:30,339    line 7072 - WARN  [org.mobicents.servlet.sip.core.session.SipApplicationSessionImpl - 1390] 
(MSS-Executor-Thread-0) 
About to release semaphore but we expected permits = 0. 
We will adjust to normal java.util.concurrent.Semaphore@2256eb0f[Permits = 1] 
app session=ba2a3a01;RestComm

**Comment by hrosa** [comment]
@RomanDS at what point is this play happening in the workflow of the app and who is the target media groups: conference's or call's?

**Comment by RomanDS** [comment]
From what I see it seems to be NC to MG connection so I think it is a play to a participant.

**Comment by hrosa** [comment]
Can you comment on the implementation of the [SdpPortManagerEvent listener](https://github.com/RestComm/RestComm-Core/blob/master/restcomm/restcomm.mscontrol.jsr309/src/main/java/org/mobicents/servlet/restcomm/mscontrol/jsr309/Jsr309CallController.java#L267)?

I believe we should add a JoinEventListener to the NC, then proceed with the join request and only proceed the JSR309CallController to an active state when the JoinEventListener receives a response.
What do you think @RomanDS ?

**Comment by RomanDS** [comment]
Hi Henrique,
Are you waiting for SDP Answer event ?

Regards,
Roman



Sent via the Samsung GALAXY S® 5, an AT&T 4G LTE smartphone
-------- Original message --------

**Comment by hrosa** [comment]
Yes, so first we do SDP negotiation and if all goes well then we do the join between NC and MG.

Seems we need a join listener attached to the MG to wait for confirmation from the Media Server, before we move to "active" state.

Can you confirm?

**Comment by RomanDS** [comment]
Hello Henrique,
You should not have to as the join is synchronous and it will block until it is complete. We noticed that the play request happened on a different thread and hence we speculated that perhaps two threads are accessing same resource.

Let me know what you think.

Regards,
Roman

**Comment by hrosa** [comment]
Hi Roman, the question here is blocking is done at JSR309/XMS side, right?
There is nothing to prevent RestComm to [move to active state](https://github.com/RestComm/RestComm-Core/blob/master/restcomm/restcomm.mscontrol.jsr309/src/main/java/org/mobicents/servlet/restcomm/mscontrol/jsr309/Jsr309CallController.java#L270) as soon as join request is issued to the driver.

Inside SdpListener we'll have:
```
networkConnection.addJoinListener(joinListener)
networkConnection.join(mediaGroup)
// do nothing else, wait for joinListener to receive confirmation
```

Then, in the new NCJoinListener:
```
if(event.successful()) {
   fsm.transition(event, active);
} else {
   fsm.transition(event, failed);
}
```
Makes sense?

**Comment by RomanDS** [comment]
Hi Henrique,
Not exactly. Join method is synchronous call. So there is no completion event there. Simply once join function completes the join is complete.

In an example bellow you seem to be thinking of using JoinInitiate() instead of Join. JoinInitiate is an asynchronous method which would generate Joined event when it completes.

What you had should have worked except that I noticed the thread ID of a play request is different from the one doing the join itself. If there was a way not to issue the play until that join method completes (the call to join gets unblocked) it should be all good.

Let me know if it is still not clear.

Best regards,
Roman

**Comment by hrosa** [comment]
I see, so this is a bit puzzling.
I don't believe for Call actor this issue can happen because as you saw, the call's connection joins the media group before it goes to active state. Call only processes Play requests when it's in active state.

Same thing for conference, as soon as conference is created the first thing we do is create the mixer and join it to the media group.

The restcomm app you're testing executes the following commands:
- > Say (call)
- > Say (call)
- > Say (call)
- > Dial Conference (plays beep, conference)

Where is the issue happening? At call level or conference level?

**Comment by RomanDS** [comment]
So, you are saying that the call state is set to active AFTER the join method completes, correct?

If that is the case that means that the JSR 309 lock does not hold… hm…

**Comment by hrosa** [comment]
Right. You can check [here](https://docs.google.com/document/d/12xDW469KpN2ggj1w86sgTYtp5gTbFEK0MVOSHQzjdKs/edit#heading=h.cv1jo6usivf5) the workflow of the Jsr309ConferenceController.
The workflow of the Jsr309CallController can be seen [here](https://docs.google.com/document/d/12xDW469KpN2ggj1w86sgTYtp5gTbFEK0MVOSHQzjdKs/edit#heading=h.t2l2urrf3ki8)

As soon as both controllers leave the initializing state, we guarantee the media group is joined to the network connection. Any play requests will only come after that.

**Comment by hrosa** [comment]
Just thinking out loud here: is it possible this issue can happen when you ask the conference to play the beep and one of the call legs is in the process of unjoining from the conference mixer? In that particular case, the call connection isn't joined to a mixer nor a media group.

It's the only scenario I can think of at the moment.

**Comment by RomanDS** [comment]
Hi Henrique,
I added John Cruz who is our JSR 309 developer as he has been working with me on these issues in a back end.

From what we see, it is a join between NC to MG. Here is the timeline which I had referenced before. Take a look at the timeline were showed before:


Timeline:

a)

09:58:30,247 line 6147 - XNC-2b99d4c75f1f4 Enter join with XMG-2b99d49565235 - Executor-Thread-0

b)

09:58:30,261 line 6283 - request to playAsync Proxy (XMG-2b99d49565235) MSS-Executor-Thread-9

c)

09:58:30,265 line 6365 - DlgcProxyHelper joinAsyncWithValidation :326 - Executor-Thread-0

d)

09:58:30,276 line 6547 - Play Starts Error reported - MSS-Executor-Thread-9

e)

09:58:30,323 line 7068 - Join returns complete - MSS-Executor-Thread-0

f)

09:58:30,339 line 7072 - WARN org.mobicents.servlet.sip.core.session.SipApplicationSessionImpl - 1390<https://github.com/RestComm/RestComm-Core/issues/MSS-Executor-Thread-0>
About to release semaphore but we expected permits = 0.
We will adjust to normal java.util.concurrent.Semaphore@2256eb0<https://github.com/java.util.concurrent.Semaphore/RestComm-Core/commit/2256eb0f>[Permits = 1]
app session=ba2a3a01;RestComm

a) is when a join gets called by an application. Note the Thread ID
b) is where a play request is executed by an application. Here note a different Thread ID
c) is where join is executed by connector and it should  be blocking the application join method
d) is where play error is reported
e) is where the join is actually completed
f) we spotted some WARN message – not sure if that has any bearing on the situation.

Please review the above.

Is there a way to enable logging for the application so we can track the state transition from before join to after join before the play? I think we need to make sure that play for sure does not get ran before join completes. Otherwise there is an issue with a lock where it sometimes does not work… I would hate to think that this is where the problem lies.

Best regards,
Roman
